### PR TITLE
Add PVC, Secret and Resource Requests

### PR DESCRIPTION
**Pull request by thoutenbos**
Adds optional configuration settings for mounting a Kubernetes Secret and/or persistentVolume in the pod.

### CODE REVIEW

**Comment by thoutenbos** [comment]
@jbguerraz any chance this could get merged in?

File: src/main/java/com/skilld/rundeck/plugin/step/kubernetes/KubernetesStep.java [hunk 1]
```diff
@@ -144,6 +148,24 @@ public void executeStep(PluginStepContext context, Map<String,Object> configurat
 			if(null != configuration.get("activeDeadlineSeconds")){
 				jobConfiguration.setActiveDeadlineSeconds(Long.valueOf(configuration.get("activeDeadlineSeconds").toString()));
 			}
+			if(null != configuration.get(PERSISTENTVOLUME)) {
```
Comment (jbguerraz) [inline src/main/java/com/skilld/rundeck/plugin/step/kubernetes/KubernetesStep.java]: This is great (to use the constant in the configuration.get call); would you mind changing the other calls to configuration.get so we would have it clean everywhere then ?

File: src/main/java/com/skilld/rundeck/plugin/step/kubernetes/KubernetesStep.java [hunk 2]
```diff
@@ -105,6 +107,8 @@ public KubernetesStep(final Framework framework) {
 		.property(PropertyUtil.select(RESTART_POLICY, "Restart policy", "The restart policy to apply to the job", true, "Never", Arrays.asList("Never", "OnFailure")))
 		.property(PropertyUtil.integer(COMPLETIONS, "Completions", "Number of pods to wait for success exit before considering the job complete", true, "1"))
 		.property(PropertyUtil.integer(PARALLELISM, "Parallelism", "Number of pods running at any instant", true, "1"))
+		.property(PropertyUtil.string(PERSISTENTVOLUME, "Persistent Volume", "The name of the PVC to use in this job in format <name>;<mountpath>", false, null))
```
Comment (jbguerraz) [inline src/main/java/com/skilld/rundeck/plugin/step/kubernetes/KubernetesStep.java]: We could use `<name>:<mountpath>;<name>:<mountpath>` and so have multiple volumes. Anyway that could be a future enhancement :) (same for secrets)

File: src/main/java/com/skilld/kubernetes/JobConfiguration.java [hunk 3]
```diff
@@ -95,6 +99,14 @@ public Integer getParallelism() {
 		return labels;
 	}
 
+	public String getPersistentVolumeName() { return persistentVolumeName; }
+
+	public String getPersistentVolumeMountPath() { return persistentVolumeMountPath; }
+
+	public String getSecretName() { return secretName; }
```
Comment (jbguerraz) [inline src/main/java/com/skilld/kubernetes/JobConfiguration.java]: we could have a single "getSecret"  to be consistent with the setter and have it returns an array of string or a string map for instance; what do you think ?

File: src/main/java/com/skilld/kubernetes/JobConfiguration.java [hunk 4]
```diff
@@ -95,6 +99,14 @@ public Integer getParallelism() {
 		return labels;
 	}
 
+	public String getPersistentVolumeName() { return persistentVolumeName; }
```
Comment (jbguerraz) [inline src/main/java/com/skilld/kubernetes/JobConfiguration.java]: we could have a single "getPersistentVolume"  to be consistent with the setter and have it returns an array of string or a string map for instance; what do you think ?

Comment (jbguerraz) [reply src/main/java/com/skilld/kubernetes/JobConfiguration.java]: \+ that way we could also use a single private variable persistentVolume

File: src/main/java/com/skilld/kubernetes/JobConfiguration.java (hunk 3 above)
Comment (jbguerraz) [reply src/main/java/com/skilld/kubernetes/JobConfiguration.java]: \+ that way we could also use a single private variable secret

**Comment by jbguerraz** [comment]
Thank you a lot for your contribution @thoutenbos !
Sorry for the delay; the team is highly busy until september.
I've added few comments, let me know :)

File: src/main/java/com/skilld/rundeck/plugin/step/kubernetes/KubernetesStep.java (hunk 2 above)
Comment (thoutenbos) [reply src/main/java/com/skilld/rundeck/plugin/step/kubernetes/KubernetesStep.java]: I'll leave this one for a future improvement :)

**Comment by thoutenbos** [comment]
Thanks for reviewing @jbguerraz !

I've processed your comments and added some more features from my local fork as well. Hope you have some time to look at it again before september :)

File: src/main/java/com/skilld/rundeck/plugin/step/kubernetes/KubernetesStep.java (hunk 2 above)
Comment (jbguerraz) [reply src/main/java/com/skilld/rundeck/plugin/step/kubernetes/KubernetesStep.java]: ++ created #7 to keep track of it

File: src/main/java/com/skilld/rundeck/plugin/step/kubernetes/KubernetesStep.java [hunk 5]
```diff
@@ -63,8 +64,10 @@
 public class KubernetesStep implements StepPlugin, Describable {
 	static Logger logger = Logger.getLogger(KubernetesStep.class);
 	private Framework framework;
-  
+
 	public static final String STEP_NAME = "kubernetes-step";
+	public static final String KUBEMASTER = "kubeMaster";
```
Comment (jbguerraz) [inline src/main/java/com/skilld/rundeck/plugin/step/kubernetes/KubernetesStep.java]: KUBE_MASTER to follow the actual style

File: src/main/java/com/skilld/rundeck/plugin/step/kubernetes/KubernetesStep.java [hunk 6]
```diff
@@ -63,8 +64,10 @@
...
+	public static final String KUBETOKEN = "kubeToken";
```
Comment (jbguerraz) [inline src/main/java/com/skilld/rundeck/plugin/step/kubernetes/KubernetesStep.java]: KUBE_TOKEN to follow the actual style

File: src/main/java/com/skilld/rundeck/plugin/step/kubernetes/KubernetesStep.java [hunk 7]
```diff
@@ -75,6 +78,10 @@
 	public static final String RESTART_POLICY = "restartPolicy";
 	public static final String COMPLETIONS = "completions";
 	public static final String PARALLELISM = "parallelism";
+	public static final String PERSISTENTVOLUME = "persistentVolume";
```
Comment (jbguerraz) [inline src/main/java/com/skilld/rundeck/plugin/step/kubernetes/KubernetesStep.java]: PERSISTENT_VOLUME to follow the actual style

File: src/main/java/com/skilld/rundeck/plugin/step/kubernetes/KubernetesStep.java [hunk 8]
```diff
@@ -75,6 +78,10 @@
...
+	public static final String SECRET = "secret";
+	public static final String RESOURCEREQUESTS = "resourceRequests";
```
Comment (jbguerraz) [inline src/main/java/com/skilld/rundeck/plugin/step/kubernetes/KubernetesStep.java]: RESOURCE_REQUESTS to follow the actual style

File: src/main/java/com/skilld/rundeck/plugin/step/kubernetes/KubernetesStep.java [hunk 9]
```diff
@@ -75,6 +78,10 @@
...
+	public static final String CLEANUP = "cleanUp";
```
Comment (jbguerraz) [inline src/main/java/com/skilld/rundeck/plugin/step/kubernetes/KubernetesStep.java]: CLEAN_UP  to follow the actual style

**Comment by jbguerraz** [comment]
Added few cosmetics comments. Other than that, LGTM.

**Comment by thoutenbos** [comment]
Updated

**Comment by thoutenbos** [comment]
I'm still finding a few bugs left and right, might be better to leave it open for a few days and I'll get the fixes in after more extensive testing.

**Comment by thoutenbos** [comment]
Have been running it stable for the past two weeks :-)

**Comment by jbguerraz** [comment]
Great! Let's get it in :) Thank you for your contribution Thijs!

**Pull request merged**